### PR TITLE
[EA] Emoji Banner May '24

### PR DESCRIPTION
**Pull request by oetherington**
Adds the new event banner ([relevant Figma](https://www.figma.com/file/qAHRGADVOGojqKpqbGMFFu/2024-Q2?node-id=685-8131&mode=dev)). This is loosly based on #8560, but there ended up being a lot more changes than I anticipated.

The implementation is a bit hacky, but it's pretty urgent so I think it's acceptable.

Screenshot 2024-05-08 at 13 27 01
Screenshot 2024-05-08 at 13 27 08
Screenshot 2024-05-08 at 13 28 12
Screenshot 2024-05-08 at 13 26 16
Screenshot 2024-05-08 at 13 26 13
Screenshot 2024-05-08 at 13 26 11
Screenshot 2024-05-08 at 13 26 00

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1207266092642074) by [Unito](https://www.unito.io)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -328,6 +329,7 @@
     "@graphql-codegen/cli": "^1.9.1",
     "@graphql-codegen/fragment-matcher": "^1.9.0",
     "@graphql-codegen/introspection": "^1.9.0",
+    "@types/cross-spawn": "^6.0.6",
```
Comment (oetherington) [inline package.json:332]: This is a dependency of `emoji-picker-element`

File: packages/lesswrong/components/Layout.tsx [hunk 2]
```diff
@@ -34,6 +34,7 @@ export const petrovBeforeTime = new DatabasePublicSetting<number>('petrov.before
 const petrovAfterTime = new DatabasePublicSetting<number>('petrov.afterTime', 0)
 
 import { Link } from '../lib/reactRouterWrapper';
+import { LoginPopoverContextProvider } from './hooks/useLoginPopoverContext';
```
Comment (oetherington) [inline packages/lesswrong/components/Layout.tsx:37]: I added this context to make it easier to trigger the login popup for new users. We can now do:
```
const currentUser = useCurrentUser();
const {onLogin} = useLoginPopoverContextProvider();
const onClick = useCallback(() => {
  if (!currentUser) {
    onLogin();
    return;
  }
  // Some action that requires being logged in
}, [currentUser, onLogin]);
```
from any react component to show the popup

File: packages/lesswrong/components/common/Header.tsx [hunk 3]
```diff
@@ -22,6 +22,9 @@ export const forumShortTitleSetting = new PublicInstanceSetting<string>('forumSe
 export const HEADER_HEIGHT = isBookUI ? 64 : 66;
 /** Height of top header on mobile. On Friendly UI sites, this is the same as the HEADER_HEIGHT */
 export const MOBILE_HEADER_HEIGHT = isBookUI ? 56 : HEADER_HEIGHT;
+export const EMOJIS_HEADER_HEIGHT = 236;
+
+const emojisInfoLink = "#"; // TODO
```
Comment (oetherington) [inline packages/lesswrong/components/common/Header.tsx:27]: TODO: Get a link to Toby's quick take once it's published

File: packages/lesswrong/components/common/Header.tsx [hunk 4]
```diff
@@ -450,9 +512,24 @@ const Header = ({
                   </Link>
                 </div>
               </Typography>
+              {showEmojisHeader && <div style={{flexGrow: 2}} />}
```
Comment (oetherington) [inline packages/lesswrong/components/common/Header.tsx:515]: This padding is needed to let us click through the section between the site title and the buttons on the right hand side

File: packages/lesswrong/components/ea-forum/auth/EALoginPopover.tsx [hunk 5]
```diff
@@ -216,12 +217,18 @@ const links = {
   privacy: "https://ev.org/ops/about/privacy-policy",
 } as const;
 
-export const EALoginPopover = ({open, setAction, isSignup, classes}: {
-  open: boolean,
-  setAction: (action: "login" | "signup" | null) => void,
-  isSignup: boolean,
+export const EALoginPopover = ({action: action_, setAction: setAction_, classes}: {
```
Comment (oetherington) [inline packages/lesswrong/components/ea-forum/auth/EALoginPopover.tsx:220]: These changes are just to use the new login popover context

File: packages/lesswrong/components/users/LoginForm.tsx [hunk 6]
```diff
@@ -260,9 +260,8 @@ const LoginFormEA = ({
 
   return (
     <Components.EALoginPopover
-      open={!!action}
```
Comment (oetherington) [inline packages/lesswrong/components/users/LoginForm.tsx:263]: For the new login context

File: packages/lesswrong/server/repos/DatabaseMetadataRepo.ts [hunk 7]
```diff
@@ -78,6 +81,56 @@ export default class DatabaseMetadataRepo extends AbstractRepo<"DatabaseMetadata
     return this.getGivingSeasonHearts(electionName);
   }
 
+  async getBannerEmojis(): Promise<BannerEmoji[]> {
```
Comment (oetherington) [inline packages/lesswrong/server/repos/DatabaseMetadataRepo.ts:84]: The giving season functions are still above as (I think) they're still used by LW, but they're sufficiently different that we need to just make some new ones - some cleanup is definitely needed here (dito for the graphql resolvers).

File: packages/lesswrong/components/ea-forum/onboarding/EAOnboardingInput.tsx [hunk 8]
```diff
@@ -24,34 +25,42 @@ export const EAOnboardingInput = ({
   value,
   setValue,
   placeholder,
+  As="input",
```
Comment (oetherington) [inline packages/lesswrong/components/ea-forum/onboarding/EAOnboardingInput.tsx:28]: We just reuse the onboarding input for the emoji popup as the styles are mostly the same. Probably we should move this input into a shared directory at some point, but that can be a different PR.

File: packages/lesswrong/components/ea-forum/EAEmojisHeader.tsx [hunk 9]
```diff
@@ -0,0 +1,584 @@
...
+import { useCurrentUser } from "../common/withUser";
+import { Link } from "../../lib/reactRouterWrapper";
+import { gql, useMutation, useQuery } from "@apollo/client";
+import ForumNoSSR from "../common/ForumNoSSR";
+import classNames from "classnames";
+import { userIsAdminOrMod } from "../../lib/vulcan-users";
+import { useLoginPopoverContext } from "../hooks/useLoginPopoverContext";
+
+if (isClient) {
+  require("emoji-picker-element");
```
Comment (oetherington) [inline packages/lesswrong/components/ea-forum/EAEmojisHeader.tsx:15]: This library provides the emoji picker as a web component that we wrap for use in react below - we shouldn't import it on the server as it relies on browser specific stuff

File: packages/lesswrong/components/ea-forum/EAEmojisHeader.tsx [hunk 10]
```diff
@@ -0,0 +1,584 @@
+import React, { FC, MouseEvent, ReactNode, createContext, useCallback, useContext, useEffect, useRef, useState } from "react";
```
Comment (oetherington) [inline packages/lesswrong/components/ea-forum/EAEmojisHeader.tsx:1]: I really dislike this style of architecture when we just define everything in a single big file, but this is quick and hacky so whatever. At least it'll make it easier to remove next week.

File: packages/lesswrong/components/ea-forum/EAEmojisHeader.tsx [hunk 11]
```diff
@@ -0,0 +1,584 @@
...
+          onChange?.(emoji);
+        }
+      }
+      picker.addEventListener("emoji-click", onChangeHandler);
+      return () => picker?.removeEventListener("emoji-click", onChangeHandler);
+    }
+  }, [ref, onChange]);
+  return (
+    // @ts-ignore
+    <emoji-picker ref={ref} />
```
Comment (oetherington) [inline packages/lesswrong/components/ea-forum/EAEmojisHeader.tsx:247]: There's probably a simple way to augment the interface to make this typecheck, but this is urgent so I didn't bother figuring it out.

File: packages/lesswrong/components/ea-forum/EAEmojisHeader.tsx [hunk 12]
```diff
@@ -0,0 +1,584 @@
...
+  const ref = useRef<HTMLElement | null>(null);
+  useEffect(() => {
+    const picker = ref.current;
+    if (picker) {
+      const styleId = "fm-emoji-style";
+      const styleElem = picker.shadowRoot?.getElementById(styleId);
+      if (!styleElem) {
+        const style = document.createElement("style");
+        style.id = styleId;
+        style.textContent = `input{font-family:"Inter";}`;
```
Comment (oetherington) [inline packages/lesswrong/components/ea-forum/EAEmojisHeader.tsx:231]: Styling the emoji picker is _really_ annoying, but possible (we may want to make more changes here).

File: packages/lesswrong/components/ea-forum/EAEmojisHeader.tsx [hunk 13]
```diff
@@ -0,0 +1,584 @@
...
+  return (
+    <>
+      <LWPopper
+        open
+        clickable
+        placement="bottom-start"
+        anchorEl={anchorElement}
+        key={`${insertPos.x},${insertPos.y}`}
+      >
+        <LWClickAwayListener onClickAway={onClickAway}>
```
Comment (oetherington) [inline packages/lesswrong/components/ea-forum/EAEmojisHeader.tsx:377]: I know there's some bug with `LWClickAwayListener` since the react 18 upgrade, but I tested it and it _seems_ to do what we want here.

File: packages/lesswrong/components/ea-forum/EAEmojisHeader.tsx [hunk 14]
```diff
@@ -0,0 +1,584 @@
...
+    LWClickAwayListener,
+  } = Components;
+  return (
+    <>
+      <LWPopper
+        open
+        clickable
+        placement="bottom-start"
+        anchorEl={anchorElement}
+        key={`${insertPos.x},${insertPos.y}`}
```
Comment (oetherington) [inline packages/lesswrong/components/ea-forum/EAEmojisHeader.tsx:375]: Force rerender the popup when the position changes.

File: packages/lesswrong/server/repos/DatabaseMetadataRepo.ts [hunk 15]
```diff
@@ -78,6 +81,56 @@ export default class DatabaseMetadataRepo extends AbstractRepo<"DatabaseMetadata
...
+  async addBannerEmoji(
+    userId: string,
+    emoji: string,
+    link: string,
+    description: string,
+    x: number,
+    y: number,
+    theta: number,
+  ): Promise<BannerEmoji[]> {
+    const t = new Date().toISOString();
```
Comment (oetherington) [inline packages/lesswrong/server/repos/DatabaseMetadataRepo.ts:110]: The time isn't currently used anywhere, but somebody mentioned wanting the times that entries were created when we did the interactive banner back in December so I added it here just in case.

**Comment by oetherington** [comment]
FYI @agnesstenlund

**Comment by agnesstenlund** [comment]
I had an early look and it's so great!! Thanks for putting this together so quickly.

Small things I found:

- [ ] Tooltip text isn't showing in light mode (guess maybe it's set to "always white")
Screenshot 2024-05-08 at 12 48 13
- [ ] You should be able to click anywhere on the popover and it wouldn't close it (I clicked on the emoji picker instead of hovering it and it created a new emoji further down)
Screenshot 2024-05-08 at 12 44 05

- [ ] We should block some areas from emojis, e.g. the edges (maybe 10px or so all around) and the text on the left. Also close to other emojis, sometimes I can get this close and it feels funny:
Screenshot 2024-05-08 at 12 48 13

- [ ] Slightly lighter color in dark mode would make it look slightly happier, something like this: #024c59

- [ ] The shadow is looking a bit harsh (especially in light mode), could we lower the shadow opacity there?
Screenshot 2024-05-08 at 12 51 31

- [ ] On a smaller screen the banner looks a bit too tall, would reduce it by as many pixels as possible while keeping the text on the left paddedd (maybe this is only 20px or so less)

Screenshot 2024-05-08 at 12 56 24

The emojis are also very slow to "follow" a new screen width, but I think this is probably fine since they do at last pop into place?
https://github.com/ForumMagnum/ForumMagnum/assets/13235378/65572c5a-c09d-4544-a05d-0181a5b416cb

**Comment by agnesstenlund** [comment]
- [ ] if not too annoying we could make "login" temporarily a darker grey in light mode

Screenshot 2024-05-08 at 13 04 24
Screenshot 2024-05-08 at 13 05 34

**Comment by oetherington** [comment]
@agnesstenlund I've fixed everything now except adding emojis too close to each other. This is difficult for a couple of reasons:

- It'd add an expensive check every time the user moves the mouse which is going to be bad for performance
- On smaller screens the emojis get squished closer together than on large screens, so I don't think there's any way to reliably fix it for all screen sizes at once

Maybe it's OK to just trust people to not put them right next to each other? Or mods could remove them if they do?

Since everything else is taken care of I'm going to open this up for review now.

Thanks!

**Comment by agnesstenlund** [comment]
thank you!! Looks perfect. I made the text smaller again since the new bigger text I asked for looked too big when the banner wasn't as tall (sorry!) 😅

**Comment by agnesstenlund** [comment]
Btw currently when I try to add an emoji it doesn't show up after I click "Add good news", but maybe that's just a local thing?

**Comment by agnesstenlund** [comment]
- [ ] also have another very small nitpick: could we make the lines around the emoji search box a lighter grey (e.g. darker in dark mode), 2px left padding and slightly less border radius 😬  (if this is how it looks out of the box and customising is finicky, then just ignore this)
Screenshot 2024-05-08 at 15 09 56

**Comment by agnesstenlund** [comment]
- [ ] oh I also think we need some shadow for the popover (ideally the same as when hovering a post item)
Screenshot 2024-05-08 at 15 16 50
Screenshot 2024-05-08 at 15 17 24

**Comment by oetherington** [comment]
@agnesstenlund Changes made :)

(And I think the bug you mentioned is a weird local thing - I can't repro)

File: packages/lesswrong/server/resolvers/eventEmojiResolvers.ts [hunk 16]
```diff
@@ -20,6 +20,9 @@ addGraphQLSchema(`
...
...{1F77F}]|[\u{1F7D5}-\u{1F7DF}]|[\u{1F7E0}-\u{1F7EB}]|[\u{1F7EC}-\u{1F7FF}]|[\u{1F80C}-\u{1F80F}]|[\u{1F848}-\u{1F84F}]|[\u{1F85A}-\u{1F85F}]|[\u{1F888}-\u{1F88F}]|[\u{1F8AE}-\u{1F8FF}]|\u{1F90C}|[\u{1F90D}-\u{1F90F}]|[\u{1F910}-\u{1F918}]|[\u{1F919}-\u{1F91E}]|\u{1F91F}|[\u{1F920}-\u{1F927}]|[\u{1F928}-\u{1F92F}]|\u{1F930}|\u{1F931}|\u{1F932}|[\u{1F933}-\u{1F93A}]|[\u{1F93C}-\u{1F93E}]|\u{1F93F}|[\u{1F940}-\u{1F945}]|[\u{1F947}-\u{1F94B}]|\u{1F94C}|[\u{1F94D}-\u{1F94F}]|[\u{1F950}-\u{1F95E}]|[\u{1F95F}-\u{1F96B}]|[\u{1F96C}-\u{1F970}]|\u{1F971}|\u{1F972}|[\u{1F973}-\u{1F976}]|[\u{1F977}-\u{1F979}]|\u{1F97A}|\u{1F97B}|[\u{1F97C}-\u{1F97F}]|[\u{1F980}-\u{1F984}]|[\u{1F985}-\u{1F991}]|[\u{1F992}-\u{1F997}]|[\u{1F998}-\u{1F9A2}]|\u{1F9A3}|\u{1F9A4}|[\u{1F9A5}-\u{1F9AA}]|[\u{1F9AB}-\u{1F9AD}]|\u{1F9AE}|\u{1F9AF}|[\u{1F9B0}-\u{1F9B9}]|[\u{1F9BA}-\u{1F9BF}]|\u{1F9C0}|\u{1F9C1}|\u{1F9C2}|[\u{1F9C3}-\u{1F9CA}]|\u{1F9CB}|\u{1F9CC}|[\u{1F9CD}-\u{1F9CF}]|[\u{1F9D0}-\u{1F9E6}]|[\u{1F9E7}-\u{1F9FF}]|[\u{1FA00}-\u{1FA6F}]|[\u{1FA70}-\u{1FA73}]|[\u{1FA74}-\u{1FA77}]|[\u{1FA78}-\u{1FA7A}]|[\u{1FA7B}-\u{1FA7F}]|[\u{1FA80}-\u{1FA82}]|[\u{1FA83}-\u{1FA8F}]|[\u{1FA90}-\u{1FA95}]|[\u{1FA96}-\u{1FFFD}]`;
```
Comment (jpaddison3) [inline packages/lesswrong/server/resolvers/eventEmojiResolvers.ts:23]: oh lawd

File: packages/lesswrong/components/ea-forum/EAEmojisHeader.tsx [hunk 17]
```diff
@@ -0,0 +1,641 @@
...
+    } else {
+      setHoverPos(null);
+    }
+  }, [normalizeCoords]);
+
+  const onMouseOut = useCallback(() => {
+    setHoverPos(null);
+  }, []);
+
+  const onClick = useCallback(async ({target, clientX, clientY}: MouseEvent) => {
```
Comment (jpaddison3) [inline packages/lesswrong/components/ea-forum/EAEmojisHeader.tsx:582]: Please add an analytics capture event here.

**Review by jpaddison3 — APPROVED** [review]
Not my most thorough review, but ASTLAC

**Review by jpaddison3 — APPROVED** [review]
